File: sub-projects/intelitor-v5.2/lib/cepaf/src/Cepaf/Observability/Fractal/KeyExpression.fs
namespace Cepaf.Observability.Fractal

open System
open System.Text.RegularExpressions

/// Zenoh-style Key Expression Engine for flexible log targeting.
/// Supports wildcards: * (single segment), ** (any path), $* (infix match)
/// STAMP Compliance: SC-LOG-009 (Key aliases pre-registered at startup)
module KeyExpression =

    // ============================================================
    // TYPES
    // ============================================================

    /// Compiled key expression for efficient matching
    type CompiledExpr = {
        /// Original key expression string
        Original: string

        /// Compiled regex pattern
        Regex: Regex

        /// Parsed segments for optimization
        Segments: string list

        /// Has single wildcard (*)
        HasWildcard: bool

        /// Has double wildcard (**)
        HasDoubleWildcard: bool

        /// Has infix wildcard ($*)
        HasInfixWildcard: bool

        /// Is exact match (no wildcards)
        IsExact: bool

        /// Compilation time for diagnostics
        CompiledAt: DateTimeOffset
    }

    /// Key expression parse result
    type ParseResult =
        | Exact of string
        | SingleWildcard of prefix: string * suffix: string
        | DoubleWildcard of prefix: string * suffix: string
        | InfixWildcard of prefix: string * pattern: string * suffix: string
        | Complex of segments: string list

    /// Selector with query parameters (Zenoh-style)
    type Selector = {
        KeyExpr: CompiledExpr
        Parameters: Map<string, string>
        Filters: Map<string, string>
    }

    // ============================================================
    // COMPILATION
    // ============================================================

    /// Escape special regex characters
    let private escapeRegex (s: string) =
        Regex.Escape(s)

    /// Compile a key expression to an optimized matcher
    let compile (expr: string) : Result<CompiledExpr, string> =
        try
            // Validate for invalid wildcard sequences
            if expr.Contains("***") then
                Error $"Invalid wildcard sequence '***' in key expression '{expr}'"
            else

            // Normalize separators (support both . and /)
            let normalized = expr.Replace(".", "/")

            // Check for wildcard types
            let hasWildcard = normalized.Contains("*") && not (normalized.Contains("**"))
            let hasDoubleWildcard = normalized.Contains("**")
            let hasInfixWildcard = normalized.Contains("$*")

            // Build regex pattern by escaping first, then replacing wildcards
            // Use placeholder tokens to avoid double-escaping
            let placeholder_dblwild = "\x00DBLWILD\x00"
            let placeholder_infixwild = "\x00INFIXWILD\x00"
            let placeholder_singlewild = "\x00SINGLEWILD\x00"

            let pattern =
                normalized
                    // Temporarily replace wildcards with placeholders
                    |> fun s -> s.Replace("$*", placeholder_infixwild)
                    |> fun s -> s.Replace("**", placeholder_dblwild)
                    |> fun s -> s.Replace("*", placeholder_singlewild)
                    // Escape all remaining special regex chars
                    |> escapeRegex
                    // Replace placeholders with regex patterns
                    |> fun s -> s.Replace(placeholder_infixwild, "([^/]*)")
                    |> fun s -> s.Replace(placeholder_singlewild, "([^/]+)")
                    // Handle ** patterns with special edge cases
                    // ** at start: match empty or any prefix with /
                    // ** at end: match empty or any suffix starting with /
                    // ** in middle: match empty or any path between slashes
                    |> fun s ->
                        // Pattern: "/**/" in middle -> "(/.*)?" followed by "/" (optionally any path, then /)
                        let s = s.Replace("/" + placeholder_dblwild + "/", "(/.*)?/")
                        // Pattern: "**/" at start -> "(.*/)?" (optionally any path ending with /)
                        let s = if s.StartsWith(placeholder_dblwild + "/") then
                                    "(.*/)?".ToString() + s.Substring(placeholder_dblwild.Length + 1)
                                else s
                        // Pattern: "/**" at end -> "(/.*)?$" (optionally / followed by anything)
                        let s = if s.EndsWith("/" + placeholder_dblwild) then
                                    s.Substring(0, s.Length - placeholder_dblwild.Length - 1) + "(/.*)?"
                                else s
                        // For any remaining standalone **, just use (.*)
                        s.Replace(placeholder_dblwild, "(.*)")

            // Parse segments for optimization
            let segments =
                normalized.Split('/')
                |> Array.toList

            // Compile regex with options for performance
            let regex = Regex(
                $"^{pattern}$",
                RegexOptions.Compiled ||| RegexOptions.Singleline
            )

            Ok {
                Original = expr
                Regex = regex
                Segments = segments
                HasWildcard = hasWildcard
                HasDoubleWildcard = hasDoubleWildcard
                HasInfixWildcard = hasInfixWildcard
                IsExact = not (hasWildcard || hasDoubleWildcard || hasInfixWildcard)
                CompiledAt = DateTimeOffset.UtcNow
            }
        with ex ->
            Error $"Failed to compile key expression '{expr}': {ex.Message}"

    /// Compile a key expression, throwing on error
    let compileOrThrow (expr: string) : CompiledExpr =
        match compile expr with
        | Ok compiled -> compiled
        | Error msg -> failwith msg

    // ============================================================
    // MATCHING
    // ============================================================

    /// Check if a key matches a compiled expression (O(1) for exact, O(n) for regex)
    let matches (compiled: CompiledExpr) (key: string) : bool =
        // Normalize key (support both . and / separators)
        let normalizedKey = key.Replace(".", "/")
        // Also normalize compiled.Original for comparison
        let normalizedOriginal = compiled.Original.Replace(".", "/")

        // Fast path: exact match
        if compiled.IsExact then
            normalizedKey = normalizedOriginal
        else
            compiled.Regex.IsMatch(normalizedKey)

    /// Check if a key matches an expression string (compiles on each call)
    let matchesExpr (expr: string) (key: string) : bool =
        match compile expr with
        | Ok compiled -> matches compiled key
        | Error _ -> false

    /// Check if two expressions could potentially match the same keys
    let intersects (a: CompiledExpr) (b: CompiledExpr) : bool =
        // If either is **, they always intersect
        if a.HasDoubleWildcard || b.HasDoubleWildcard then
            true
        // If both are exact, they intersect only if equal
        elif a.IsExact && b.IsExact then
            a.Original = b.Original
        // Otherwise, check if one could match the other
        else
            matches a b.Original || matches b a.Original

    // ============================================================
    // SELECTORS (Key Expression + Query Parameters)
    // ============================================================

    /// Parse a selector string (key_expr?param=value&filter.key=value)
    let parseSelector (selectorStr: string) : Result<Selector, string> =
        try
            let parts = selectorStr.Split('?')
            let keyExprStr = parts.[0]

            let parameters, filters =
                if parts.Length > 1 then
                    parts.[1].Split('&')
                    |> Array.fold (fun (pars, filts) param ->
                        let kvp = param.Split('=')
                        if kvp.Length = 2 then
                            let key = kvp.[0]
                            let value = kvp.[1]
                            if key.StartsWith("filter.") then
                                (pars, filts |> Map.add (key.Substring(7)) value)
                            else
                                (pars |> Map.add key value, filts)
                        else
                            (pars, filts)
                    ) (Map.empty, Map.empty)
                else
                    (Map.empty, Map.empty)

            match compile keyExprStr with
            | Ok compiled ->
                Ok {
                    KeyExpr = compiled
                    Parameters = parameters
                    Filters = filters
                }
            | Error msg -> Error msg
        with ex ->
            Error $"Failed to parse selector '{selectorStr}': {ex.Message}"

    /// Extract query parameters from selector
    let getParameter (selector: Selector) (key: string) : string option =
        selector.Parameters |> Map.tryFind key

    /// Extract filter from selector
    let getFilter (selector: Selector) (key: string) : string option =
        selector.Filters |> Map.tryFind key

    // ============================================================
    // KEY BUILDING
    // ============================================================

    /// Build a key from module and function
    let buildKey (moduleName: string) (functionName: string) : string =
        $"{moduleName}/{functionName}"

    /// Build a key from module, function, and event type
    let buildKeyWithEvent (moduleName: string) (functionName: string) (eventType: string) : string =
        $"{moduleName}/{functionName}/{eventType}"

    /// Extract module name from key
    let extractModule (key: string) : string option =
        let normalized = key.Replace(".", "/")
        let parts = normalized.Split('/')
        if parts.Length >= 1 then Some parts.[0]
        else None

    /// Extract function name from key
    let extractFunction (key: string) : string option =
        let normalized = key.Replace(".", "/")
        let parts = normalized.Split('/')
        if parts.Length >= 2 then Some parts.[parts.Length - 1]
        else None

    // ============================================================
    // PATTERN EXAMPLES
    // ============================================================

    /// Common key expression patterns for reference
    module Patterns =
        /// Match all events in a module
        let allInModule moduleName = $"{moduleName}/**"

        /// Match all create events anywhere
        let allCreate = "**/create"

        /// Match all errors anywhere
        let allErrors = "**/error"

        /// Match specific function in any module
        let functionInAny funcName = $"**/{funcName}"

        /// Match any Handler suffix
        let anyHandler = "**/$*Handler"

        /// Match Cortex cognitive events
        let cortexCognitive = "Indrajaal/Cortex/**"

        /// Match security audit events
        let securityAudit = "Indrajaal/Security/**"

        /// Match all alarms
        let allAlarms = "Indrajaal/Alarms/**"

    // ============================================================
    // VALIDATION
    // ============================================================

    /// Validate a key expression syntax
    let validate (expr: string) : Result<unit, string list> =
        let errors = ResizeArray<string>()

        // Check for empty
        if String.IsNullOrWhiteSpace(expr) then
            errors.Add("Key expression cannot be empty")

        // Check for invalid characters
        let invalidChars = [| '<'; '>'; '|'; '\\'; '"'; '''|]
        for c in invalidChars do
            if expr.Contains(c) then
                errors.Add($"Invalid character '{c}' in key expression")

        // Check for adjacent wildcards
        if expr.Contains("***") then
            errors.Add("Invalid wildcard sequence '***'")

        // Check for leading/trailing slashes
        if expr.StartsWith("/") || expr.EndsWith("/") then
            errors.Add("Key expression should not start or end with '/'")

        // Try compilation as final check
        match compile expr with
        | Error msg -> errors.Add(msg)
        | Ok _ -> ()

        if errors.Count = 0 then Ok ()
        else Error (errors |> Seq.toList)

    /// Check if an expression is valid
    let isValid (expr: string) : bool =
        match validate expr with
        | Ok _ -> true
        | Error _ -> false

File: intelitor-v5.2/lib/indrajaal/observability/fractal/key_expression.ex
defmodule Indrajaal.Observability.Fractal.KeyExpression do
  @moduledoc """
  Zenoh-style Key Expression Engine for Fractal Logging.

  Provides flexible log targeting with wildcards:
  - `*` - Match exactly one path segment
  - `**` - Match zero or more path segments
  - `$*` - Infix match within a segment

  ## Examples

      # Exact match
      matches?("Indrajaal/Alarms/create", "Indrajaal/Alarms/create")  # true

      # Single wildcard
      matches?("Indrajaal/*/create", "Indrajaal/Alarms/create")  # true
      matches?("Indrajaal/*/create", "Indrajaal/A/B/create")  # false

      # Double wildcard
      matches?("Indrajaal/**", "Indrajaal/A/B/C/D")  # true
      matches?("**/error", "Any/Path/To/error")  # true

      # Infix wildcard
      matches?("**/$*Handler", "Module/AlarmHandler")  # true

  ## STAMP Compliance

  - SC-LOG-009: Key aliases pre-registered at startup
  """

  @type compiled_expr :: %{
          original: String.t(),
          regex: Regex.t(),
          segments: [String.t()],
          has_wildcard: boolean(),
          has_double_wildcard: boolean(),
          has_infix_wildcard: boolean(),
          is_exact: boolean()
        }

  # ============================================================
  # COMPILATION
  # ============================================================

  @doc """
  Compile a key expression to an optimized matcher.
  """
  @spec compile(String.t()) :: {:ok, compiled_expr()} | {:error, String.t()}
  def compile(expr) when is_binary(expr) do
    # Normalize separators
    normalized = String.replace(expr, ".", "/")

    has_wildcard = String.contains?(normalized, "*") and not String.contains?(normalized, "**")
    has_double_wildcard = String.contains?(normalized, "**")
    has_infix_wildcard = String.contains?(normalized, "$*")
    is_exact = not (has_wildcard or has_double_wildcard or has_infix_wildcard)

    # Build regex pattern
    # Use placeholders to protect wildcard patterns during escaping
    pattern =
      normalized
      |> String.replace("$*", "\x00INFIX\x00")
      |> String.replace("**", "\x00DOUBLE\x00")
      |> String.replace(~r/(?<!\*)(\*)(?!\*)/, "\x00SINGLE\x00")
      |> escape_regex_special()
      |> String.replace("\x00INFIX\x00", "([^/]*)")
      # Handle ** followed by / - makes both content AND slash optional together
      |> String.replace("\x00DOUBLE\x00/", "(.*/)?")
      # Handle ** at end or standalone - matches anything after
      |> String.replace("\x00DOUBLE\x00", "(.+)?")
      |> String.replace("\x00SINGLE\x00", "([^/]+)")

    case Regex.compile("^#{pattern}$") do
      {:ok, regex} ->
        {:ok,
         %{
           original: expr,
           regex: regex,
           segments: String.split(normalized, "/"),
           has_wildcard: has_wildcard,
           has_double_wildcard: has_double_wildcard,
           has_infix_wildcard: has_infix_wildcard,
           is_exact: is_exact
         }}

      {:error, reason} ->
        {:error, "Failed to compile key expression: #{inspect(reason)}"}
    end
  end

  @doc """
  Compile a key expression, raising on error.
  """
  @spec compile!(String.t()) :: compiled_expr()
  def compile!(expr) do
    case compile(expr) do
      {:ok, compiled} -> compiled
      {:error, reason} -> raise ArgumentError, reason
    end
  end

  # ============================================================
  # MATCHING
  # ============================================================

  @doc """
  Check if a key matches a compiled expression.
  """
  @spec matches?(compiled_expr(), String.t()) :: boolean()
  def matches?(%{is_exact: true, original: original}, key) do
    normalized_key = String.replace(key, ".", "/")
    normalized_original = String.replace(original, ".", "/")
    normalized_key == normalized_original
  end

  def matches?(%{regex: regex}, key) do
    normalized_key = String.replace(key, ".", "/")
    Regex.match?(regex, normalized_key)
  end

  @spec matches?(String.t(), String.t()) :: boolean()
  def matches?(expr, key) when is_binary(expr) do
    case compile(expr) do
      {:ok, compiled} -> matches?(compiled, key)
      {:error, _} -> false
    end
  end

  @doc """
  Check if two expressions could match the same keys.
  """
  @spec intersects?(compiled_expr(), compiled_expr()) :: boolean()
  def intersects?(a, b) do
    cond do
      a.has_double_wildcard or b.has_double_wildcard -> true
      a.is_exact and b.is_exact -> a.original == b.original
      true -> matches?(a, b.original) or matches?(b, a.original)
    end
  end

  # ============================================================
  # KEY BUILDING
  # ============================================================

  @doc """
  Build a key from module and function.
  """
  @spec build_key(atom() | String.t(), atom() | String.t()) :: String.t()
  def build_key(module, function) do
    module_str = module |> to_string() |> String.replace("Elixir.", "")
    "#{module_str}/#{function}"
  end

  @doc """
  Build a key with event type.
  """
  @spec build_key(atom() | String.t(), atom() | String.t(), atom() | String.t()) :: String.t()
  def build_key(module, function, event_type) do
    "#{build_key(module, function)}/#{event_type}"
  end

  @doc """
  Extract the module name from a key.
  """
  @spec extract_module(String.t()) :: String.t() | nil
  def extract_module(key) do
    key
    |> String.replace(".", "/")
    |> String.split("/")
    |> List.first()
  end

  @doc """
  Extract the function name from a key.
  """
  @spec extract_function(String.t()) :: String.t() | nil
  def extract_function(key) do
    key
    |> String.replace(".", "/")
    |> String.split("/")
    |> List.last()
  end

  # ============================================================
  # VALIDATION
  # ============================================================

  @doc """
  Validate a key expression syntax.
  """
  @spec validate(String.t()) :: :ok | {:error, [String.t()]}
  def validate(expr) do
    errors = []

    errors =
      if String.trim(expr) == "" do
        ["Key expression cannot be empty" | errors]
      else
        errors
      end

    errors =
      if Regex.match?(~r/[<>|\\"]/, expr) do
        ["Invalid characters in key expression" | errors]
      else
        errors
      end

    errors =
      if String.contains?(expr, "***") do
        ["Invalid wildcard sequence '***'" | errors]
      else
        errors
      end

    errors =
      if String.starts_with?(expr, "/") or String.ends_with?(expr, "/") do
        ["Key expression should not start or end with '/'" | errors]
      else
        errors
      end

    case compile(expr) do
      {:ok, _} ->
        if errors == [], do: :ok, else: {:error, Enum.reverse(errors)}

      {:error, reason} ->
        {:error, Enum.reverse([reason | errors])}
    end
  end

  @doc """
  Check if an expression is valid.
  """
  @spec valid?(String.t()) :: boolean()
  def valid?(expr) do
    validate(expr) == :ok
  end

  # ============================================================
  # COMMON PATTERNS
  # ============================================================

  @doc """
  Common key expression patterns.
  """
  def patterns do
    %{
      all_in_module: fn module -> "#{module}/**" end,
      all_create: "**/create",
      all_errors: "**/error",
      function_in_any: fn func -> "**/#{func}" end,
      any_handler: "**/$*Handler",
      cortex_cognitive: "Indrajaal/Cortex/**",
      security_audit: "Indrajaal/Security/**",
      all_alarms: "Indrajaal/Alarms/**"
    }
  end

  # ============================================================
  # PRIVATE
  # ============================================================

  defp escape_regex_special(str) do
    # Escape special regex chars (wildcards are already replaced with placeholders)
    String.replace(str, ~r/([.+?^${}()|[\]\\])/, "\\\\\\1")
  end
end
